src/miner.cpp

```diff
@@ -662,31 +662,171 @@ bool SignStakeBlock(CBlock &block, CKey &key, vector<const CWalletTx*> &StakeInp
+            return;
+        if(neural_response.length() != 32)
+            return;
+        if("d41d8cd98f00b204e9800998ecf8427e"==neural_response)
+            return;
+        const std::string logprefix = "supercfwd.HashResponseHook: from "+fromNode->addrName+" hash "+neural_response;
+
+        if(neural_response!=sCacheHash)
+        {
+            double popularity = 0;
+            const std::string consensus_hash = GetNeuralNetworkSupermajorityHash(popularity);
+
+            if(neural_response==consensus_hash)
+            {
+                if(fDebug) LogPrintf("%s requesting contract data",logprefix);
+                fromNode->PushMessage(/*command*/ "neural", /*subcommand*/ std::string("quorum"), /*reqid*/std::string("supercfwd.hrh"));
+            }
+            else
+            {
+                if(fDebug) LogPrintf("%s not matching consensus",logprefix);
+                //TODO: try another peer faster
+            }
+        }
+        else if(fDebug) LogPrintf("%s already cached",logprefix);
+    }
+
+    void QuorumResponseHook(CNode* fromNode, const std::string& neural_response)
+    {
+        assert(fromNode);
+        const auto resp_length= neural_response.length();
+
+        if(fEnable && resp_length >= 10)
+        {
+            const std::string rcvd_contract= UnpackBinarySuperblock(std::move(neural_response));
+            const std::string rcvd_hash = GetQuorumHash(rcvd_contract);
+            const std::string logprefix = "supercfwd.QuorumResponseHook: from "+fromNode->addrName + " hash "+rcvd_hash;
+
+            if(rcvd_hash!=sCacheHash)
+            {
+                double popularity = 0;
+                const std::string consensus_hash = GetNeuralNetworkSupermajorityHash(popularity);
+
+                if(rcvd_hash==consensus_hash)
+                {
+                    LogPrintf("%s good contract save",logprefix);
+                    sBinContract= PackBinarySuperblock(std::move(rcvd_contract));
+                    sCacheHash= std::move(rcvd_hash);
+
+                    if(!fNoListen)
+                        SendOutRcvdHash();
+                }
+                else
+                {
+                    if(fDebug) LogPrintf("%s not matching consensus, (size %d)",logprefix,resp_length);
+                    //TODO: try another peer faster
+                }
+            }
+            else if(fDebug) LogPrintf("%s already cached",logprefix);
+        }
+        //else if(fDebug10) LogPrintf("%s invalid data",logprefix);
     }
+    void SendResponse(CNode* fromNode, const std::string& req_hash)
+    {
+        const std::string nn_hash(NN::GetNeuralHash());
+        const bool& fDebug10= fDebug; //temporary
+        if(req_hash==sCacheHash)
+        {
+            if(fDebug10) LogPrintf("supercfwd.SendResponse: %s requested %s, sending forwarded binary contract (size %d)",fromNode->addrName,req_hash,sBinContract.length());
+            fromNode->PushMessage("neural", std::string("supercfwdr"),
```

Comment:
Outside of this scope, but what do you say we make these message constants later?

`const std::string SUPERBLOCK_FWD_RESP("supercfwdr");`

Reply:
Absolutely!
The "message type" (first string in message, saying what it is) should even be 8-bit integer constant.